File: js/stage3/S3_PIPE_03.js
// S3_PIPE_03 — String normalization pipeline
const pipe =
  (...fns) =>
  (value) =>
    fns.reduce((acc, fn) => fn(acc), value);
const normalize = pipe(
  (s) => s.trim(),
  (s) => s.toLowerCase(),
  (s) => s.replace(/\s+/g, " ")
);
// Test
console.log(normalize("  HeLLo    WoRLd   "));

File: Python/stage3/S3_PIPE_03.py
# S3_PIPE_03 — String normalization pipeline

import re

def pipe(*fns):

    def runner(x):

        for fn in fns:

            x = fn(x)

        return x

    return runner


normalize = pipe(

    lambda s: s.strip(),

    lambda s: s.lower(),

    lambda s: re.sub(r"\s+", " ", s),

)

print(normalize(" Ala   Ma    Kota "))
 